src/frontend/editor.ts

```diff
@@ -15,4 +15,8 @@ editor.setOptions({
     enableLiveAutocompletion: true
 });
 
-onChange('source', () => editor.getSession().setValue(state.source || codeElement.innerText));
+onChange('source', () => {
+    if (editor.getSession().getValue() !== state.source) {
```

Comment:
One nitpick (optional): could you structure it so that the error case is an early return from the function?

```javascript
onChange('source', () => {
    if (editor.getSession().getValue() === state.source) return;
    editor.getSession().setValue(state.source || codeElement.innerText);
});
```

Reply:
This is a huge nit tho, I'm just a big fan of guard statements. sorry for being so picky 😂